### PR TITLE
Distributed data parallelism

### LINKED ISSUE #2: Run with distributed parallelism
Add `torch.nn.parallel.DistributedDataParallel`. Current plan:

- [x] Replace `DataParallel` calls with `DDP`
~~Inherit from `Dataset` for data~~
- [ ] Wrap data with `DistributedSampler`
- [ ] Add 2 rank test

### PR DESCRIPTION
**Pull request by allaffa**
@pzhanggit @streeve 
I assign you as reviewers for my pull request that runs DistributedDataParallelismn on GCNN
I tested the code also on my DGX machine to make sure that the distributed processes take advantage of the GPUs when needed. 

Closes #2

### CODE REVIEW

File: utils/utils.py
Comment (streeve) [inline utils/utils.py:91]: This is an annoying API difference

File: utils/utils.py
Comment (streeve) [inline utils/utils.py:101]: Importing directly could shorten this quite a bit

File: utils/utils.py
Comment (streeve) [inline utils/utils.py:256]: So you can still wrap the existing data with `DistributedSampler` without inheriting?

File: utils/utils.py
Comment (streeve) [inline utils/utils.py:288]: May as well rename these `trainset` and `testset` (unless you're worried about backwards compatibility)

File: utils/utils.py
Comment (streeve) [inline utils/utils.py:254]: This section for splitting into test/train/validation for parallel/serial should be a separate function called by both `split_dataset` and `combine_and_split_datasets`

File: utils/models_setup.py
Comment (streeve) [inline utils/models_setup.py:43]: I would extract this the same way as initialization, so maybe a separate function called within `setup`?

File: run_config_input.py
Comment (streeve) [inline run_config_input.py:226]: This breaks on CPU. And in general, shouldn't we call `get_gpu` (`get_device`) here instead of doing this directly?

File: utils/models_setup.py
Comment (streeve) [inline utils/models_setup.py:50]: I think that calculation should be done inside `get_gpu`

**Review by streeve — COMMENTED** [review]
Just a first round of review to try to understand and clean up the interfaces. 

Having a simple class for parallel details that we pass around could simplify a lot (it could hold a `bool` for whether we're running in parallel, `world_size`, etc.)

**Comment by streeve** [comment]
Also, per #2 we'll want a test that runs in parallel before we merge

File: distributed_data_parallelism_main.py
Comment (allaffa) [reply distributed_data_parallelism_main.py:8]: Yes, but you cannot allow the user to use an interactive input selection if you want to to run things in parallel

File: run_config_input.py
Comment (allaffa) [reply run_config_input.py]: @streeve I push ed changes to take care of this.

File: utils/models_setup.py
Comment (allaffa) [reply utils/models_setup.py]: removed

File: utils/models_setup.py
Comment (allaffa) [reply utils/models_setup.py:28]: This is needed to grab the "device" with a specific index. The function `torch.device()` takes as an input argument 
a string like 'cuda:number'

File: utils/models_setup.py
Comment (allaffa) [reply utils/models_setup.py:49]: get_gpu returns the torch.device object associated with a specific number ID, assuming that the number ID is in the liste of the available CUDA devices, otherwise it returns an error. 
In view of a potential extension of the code where a single neural network does not fit on a single GPU, what you propose may be limiting.

File: utils/utils.py
Comment (allaffa) [reply utils/utils.py:290]: actually there are 3 partitioning performed on 2 datasets

dataset1 is split into training and validation. dataset 2 becomes the testset. 
So renaming dataset1 as trainingset may be confusing

File: run_config_input.py
Comment (allaffa) [reply run_config_input.py]: Changed as suggested

File: utils/utils.py
Comment (allaffa) [reply utils/utils.py:242]: changed as suggested

**Comment by streeve** [comment]
@allaffa I updated to run on cpu as well and added an mpi test. This should pass when we rebase on #18

**Comment by allaffa** [comment]
@streeve Is it normal that the tests do not pass? i think you did not run black before pushing the last changes.

**Comment by streeve** [comment]
One question is a good way to test correctness for these runs - right now it just makes sure the code runs without errors, but maybe we leave that to the next PR

**Review by streeve — APPROVED** [review]

**Comment by streeve** [comment]
Also, @pzhanggit there are "approve"/"request changes" options in the code changes tab

**Comment by streeve** [comment]
@allaffa merge when you like

**Comment by allaffa** [comment]
@streeve done

**Pull request merged**
